gatey_sdk/buffer.py

```diff
@@ -0,0 +1,235 @@
+"""
+    Events `buffer` class, that does storing events (as raw data)
+    and handles them passing to the transport.
+
+    Something like `proxy` that can just immediatly send event directly to the transport
+    or can buffer them and wait for next requirement in their send right now.
+
+    Includes handling exit signal to not loss any events,
+    and flush thread if required (will refresh send every `N` time).
+"""
+import atexit
+from typing import Dict, List, Any, Optional, Callable
+from time import sleep
+from threading import Thread
+
+from gatey_sdk.transports.base import BaseTransport
+from gatey_sdk.consts import (
+    DEFAULT_EVENTS_BUFFER_FLUSH_EVERY,
+    EVENTS_BUFFER_FLUSHER_THREAD_NAME,
+)
+
+
+class _EventsBufferFlusher:
+    """
+    Events `buffer` flusher class, does handling auto refresh (every time, at exit).
+
+    Includes handling exit signal to not loss any events,
+    and flush thread if required (will refresh send every `N` time).
+    """
+
+    # Thread that is used for periodically flushing events to be passed to transport.
+    flush_thread: Optional[Thread] = None
+
+    # Settings.
+    flush_every: float = DEFAULT_EVENTS_BUFFER_FLUSH_EVERY
+    on_flush: Callable[[], Any]
+
+    def __init__(
+        self,
+        on_flush: Callable[[], Any],
+        *,
+        flush_every: float = DEFAULT_EVENTS_BUFFER_FLUSH_EVERY
+    ):
+        """
+        :param on_flush: Callable that will be called on flush request.
+        :param flush_every: Time in seconds for refreshing and flushing events (passing to the transport), (left 0 to disable)
+        """
+
+        # Settings.
+        self.on_flush = on_flush
+        self.flush_every = float(flush_every)
+
+        # Setup.
+        self.ensure_running_thread()
+        self.bind_system_exit_hook()
+
+    def ensure_running_thread(self) -> Thread:
+        """
+        Runs buffer flush thread if it is not running, and returns thread.
+        Flush thread is used to send events buffer after some time, not causing to wait core application
+        for terminate or new events that will trigger bulk sending (buffer flush).
+        :returns Thread: Flush thread.
+        """
+
+        if isinstance(self.flush_thread, Thread) and self.flush_thread.is_alive():
+            # If thread is currently alive - there is no need to create new thread by removing old reference.
+            return self.flush_thread
+        return self._spawn_new_thread()
+
+    def flush_thread_target(self) -> None:
+        """
+        Thread target for events buffer flusher.
+        """
+        while True:
```

Comment:
Use `threading.Timer` instead of `threading.Thread` and remove `while True` and `time.sleep`